pkg/argocd/update.go

```diff
@@ -561,76 +565,118 @@ func mergeKustomizeOverride(t *kustomizeOverride, o *kustomizeOverride) {
 	}
 }
 
-// Check if a key exists in a MapSlice and return its index and value
-func findHelmValuesKey(m yaml.MapSlice, key string) (int, bool) {
-	for i, item := range m {
-		if item.Key == key {
-			return i, true
+func splitKeys(input string) []string {
+	// Regular expression to match quoted and unquoted segments
+	re := regexp.MustCompile(`'([^']*)'|"([^"]*)"|([^.".]+)`)
+	matches := re.FindAllStringSubmatch(input, -1)
+
+	var result []string
+	for _, match := range matches {
+		if match[1] != "" { // Single-quoted string
+			result = append(result, match[1])
+		} else if match[2] != "" { // Double-quoted string
+			result = append(result, match[2])
+		} else if match[3] != "" { // Unquoted segment
+			result = append(result, match[3])
 		}
 	}
-	return -1, false
+
+	return result
 }
 
 // set value of the parameter passed from the annotations.
-func setHelmValue(currentValues *yaml.MapSlice, key string, value interface{}) error {
-	// Check if the full key exists
-	if idx, found := findHelmValuesKey(*currentValues, key); found {
-		(*currentValues)[idx].Value = value
-		return nil
```

Comment:
the current behaviro is, if the full key (e.g., `images.nginx.name`) already exists in the target helm file, take this existing key and update its value. Otherwise, assume the users want nested parameter names. I think this should be perserved when the annotation value `helm.image-name` and `helm.image-tag` are not explicitly quoted.

Even when the 2 annotation values are explicitly quoted, it's still worth considering whether we should honor the existing key format (flat dotted or nested) over the application annotation.